internal/weldr/api.go

```diff
@@ -91,18 +93,114 @@ func (api *API) systemRepoNames() (names []string) {
+	distroRepos[distro.Name()][arch.Name()] = repos
+	distros, _ := distroregistry.New(distro)
+
 	api := &API{
 		store:           store,
 		workers:         workers,
-		rpmmd:           rpmmd,
-		arch:            arch,
-		distro:          distro,
-		repos:           repos,
+		rpmmd:           rpm,
 		logger:          logger,
 		compatOutputDir: compatOutputDir,
+		hostDistroName:  distro.Name(),
+		distros:         distros,
+		distroRepos:     distroRepos,
+	}
+	return setupRouter(api)
+}
+
+// TODO this naming needs to be sorted out in the distro/repo code someplace
+func mangleName(name string, beta, stream bool) string {
+	if name == "rhel-84" {
+		name = "rhel-8"
+	}
+	if beta {
+		name += "-beta"
+	}
+
+	// override repository for centos stream, remove when CentOS 8 is EOL
+	if stream && name == "centos-8" {
+		name = "centos-stream-8"
+	}
+	return name
+}
+
+func New(repoPaths []string, stateDir string, rpm rpmmd.RPMMD,
+	distros *distroregistry.Registry, logger *log.Logger,
+	workers *worker.Server) (*API, error) {
+
+	hostDistro, beta, isStream, err := distros.FromHost()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: refactor to be more generic
+	name := mangleName(hostDistro.Name(), beta, isStream)
+
+	// Load the repositories for all the supported distros
+	distroRepos := make(map[string]map[string][]rpmmd.RepoConfig, len(distros.List()))
+	for _, distro := range distros.List() {
+		// TODO How to mangle this for non-host distro?
+		repoName := mangleName(distro, false, false)
+		repo, err := rpmmd.LoadRepositories(repoPaths, repoName)
+		if err != nil {
+			return nil, fmt.Errorf("Error loading repositories for %s/%s: %v", distro, repoName, err)
+		}
+		distroRepos[distro] = repo
 	}
 
+	hostArch, err := hostDistro.GetArch(common.CurrentArch())
+	if err != nil {
+		return nil, err
+	}
+	store := store.New(&stateDir, hostArch, logger)
+	compatOutputDir := path.Join(stateDir, "outputs")
+
+	api := &API{
+		store:           store,
+		workers:         workers,
+		rpmmd:           rpm,
+		logger:          logger,
+		compatOutputDir: compatOutputDir,
+		hostDistroName:  name,
+		distros:         distros,
+		distroRepos:     distroRepos,
+	}
+	return setupRouter(api), nil
+}
+
+// arch returns the Arch struct for the selected distro
+func (api *API) arch(distro string) (distro.Arch, error) {
+	// Currently we only support the same arch as the host
+	archName := common.CurrentArch()
+	d := api.distros.GetDistro(distro)
+	if d == nil {
+		return nil, fmt.Errorf("Unsuppoered distro: %s", distro)
+	}
+	return d.GetArch(archName)
+}
+
+// platformID returns the module platform string for the selected distro
+func (api *API) platformID(distro string) (string, error) {
+	d := api.distros.GetDistro(distro)
+	if d == nil {
+		return "", fmt.Errorf("Unsupported distro: %s", distro)
+	}
+	return d.ModulePlatformID(), nil
+}
+
+func (api *API) hostRepos() []rpmmd.RepoConfig {
+	return api.distroRepos[api.hostDistroName][common.CurrentArch()]
```

Comment:
After reading the code, I think that this won't work if the `hostDistroName` will be `centos-stream-8` or `rhel-8-beta`. the `distroRepos` map does not reflect, that the same distro definition may be GA/Beta or non-Stream/Stream. In the end, the difference is IMO mainly in the repositories used for building the image, but this shortcut in distro definitions is actually making the problem here harder...

Reply:
Ah, now I see what you mean. It would be cleaner if we didn't mangle names and every variant had it's own distro name. It's possible someone may want to build for beta vs. stream vs. a specific release.

Reply:
Yeah, I fully agree. But given our current options of how distros are defined internally in composer, this would create additional copies for each distro (read as copies of Go code). I do believe that we should not do it before we have a better way of defining distros / arches / image types, including inheritance...

Reply:
Will that kind of re-structuring be part of #1387? It's probably best to do that first and then I can rebase on it and adjust accordingly.

Reply:
@bcl no, the restructuring of how are Distros defined won't be part of https://github.com/osbuild/osbuild-composer/pull/1387 but I included a change which will ensure that the `distros.FromHost()` returns a `distro.Distro` instance, which `Name()` returns the correctly "mangled" name. This is implemented by https://github.com/osbuild/osbuild-composer/pull/1387/commits/06077e5df2c54de5286d4fb26146ad6ed15c7a9b . I invite you to provide feedback if you have any.. :wink:

But I do believe that the distroregistry API and how it is used should not change after https://github.com/osbuild/osbuild-composer/pull/1387 is merged, even when we restructure how are Distros defined. I think that the abstractions should be already good enough to hide such work from Weldr's point of view.